### PR TITLE
Add basic peer to peer functionality

### PR DESCRIPTION
**Pull request by ErisMik**

### CODE REVIEW

File: bookchain-miner/src/http/dtos/dtos.hpp [hunk 1]
```diff
@@ -9,27 +8,14 @@ namespace bookchain::http {
...
-class BasicMessageDto : public oatpp::data::mapping::type::Object {
-    DTO_INIT(BasicMessageDto, Object)
+class HelloMessageDto : public oatpp::data::mapping::type::Object {
+    DTO_INIT(HelloMessageDto, Object)
 
-    DTO_FIELD(Int32, statusCode);
-    DTO_FIELD(String, message);
-};
-
-class BlockDto : public oatpp::data::mapping::type::Object {
```
Comment (MuchToKnow) [inline bookchain-miner/src/http/dtos/dtos.hpp:19]: Do we even need this file now that this is moved to blockdtos.hpp?

File: bookchain-miner/src/http/dtos/peerdtos.hpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+#include <oatpp/core/macro/codegen.hpp>
+#include <oatpp/parser/json/mapping/ObjectMapper.hpp>
+
+namespace bookchain::http {
+
+/* Begin DTO code-generation */
+#include OATPP_CODEGEN_BEGIN(DTO)
+
+class PeerLinkDto : public oatpp::data::mapping::type::Object {
```
Comment (MuchToKnow) [inline bookchain-miner/src/http/dtos/peerdtos.hpp:11]: Can you explain the logic behind separating peers into 2 separate DTOs?

File: bookchain-miner/src/main.cpp [hunk 3]
```diff
@@ -43,5 +46,35 @@ int main(int /*argc*/, const char* /*argv*/[]) {
     std::cout << bookchain::utils::hexifystring(fifthBlock.blockHash()) << std::endl;
     std::cout << bookchain::utils::hexifystring(bloockchain.bloock(testBlockHeight).blockHash()) << std::endl;
 
-    bookchain::http::startMinerHttpServer();
+    std::cout << "Miner thread stopped" << std::endl;
+
+    std::cout << "Looping!" << std::endl;
+    bookchain::PeersList peersList;
+    while (true) {
```
Comment (MuchToKnow) [inline bookchain-miner/src/main.cpp:53]: For now this is just testing and peering would be in a separate thread from the miner (probably the Node thread) correct?

File: bookchain-miner/src/queue.hpp [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    std::queue<T> _queue;
+    std::mutex _queueMutex;
+};
+
+template <typename T>
+using sharedTSQueue = std::shared_ptr<ThreadsafeQueue<T>>;
+
+template <typename T>
+void ThreadsafeQueue<T>::push(T& element) {
+    std::lock_guard<std::mutex> guard(this->_queueMutex);
```
Comment (MuchToKnow) [inline bookchain-miner/src/queue.hpp:27]: Til you don't need to manually release mutexes in c++ :+1:

This language is so different in industry vs academics lol

File: bookchain-miner/src/storage.cpp [hunk 5]
```diff
@@ -18,7 +18,7 @@ void purgeChain(const std::string& filename) {
     blockchainFile.close();
 }
 
-void dumpChain(std::vector<Bloock> bloockChain, const std::string& filename) {
+void dumpChain(std::vector<Bloock>& bloockChain, const std::string& filename) {
```
Comment (MuchToKnow) [inline bookchain-miner/src/storage.cpp:21]: Do we need a function to load a blockchain from file?

**Comment by MuchToKnow** [comment]
One more general comment. I didn't see the client side of peering (I.e pinging all ips looking for initial peers)?  Did I miss that somewhere and if so can you point it out?

File: bookchain-miner/src/http/dtos/dtos.hpp (hunk 1 above)
Comment (ErisMik) [reply bookchain-miner/src/http/dtos/dtos.hpp:19]: I was keeping it for the general dtos, I have a feeling we will have things to put here. Also good for testing for now.

File: bookchain-miner/src/http/dtos/peerdtos.hpp (hunk 2 above)
Comment (ErisMik) [reply bookchain-miner/src/http/dtos/peerdtos.hpp:11]: One is for linking one is the actual data. So the PeerLinkDto will be used to syncing up /linking two nodes, whereas the PeerDto represents an actual peer object and is used to transmit peer data (for the whole peers list)

File: bookchain-miner/src/main.cpp (hunk 3 above)
Comment (ErisMik) [reply bookchain-miner/src/main.cpp:53]: Absolutely! This is literally just a smoke test function right now.

File: bookchain-miner/src/queue.hpp (hunk 4 above)
Comment (ErisMik) [reply bookchain-miner/src/queue.hpp:27]: I think it's cause in academics they typically use C or like C style C++. Once you start to actual use the modern C++ features that separate it from C things become much nicer.

File: bookchain-miner/src/storage.cpp (hunk 5 above)
Comment (ErisMik) [reply bookchain-miner/src/storage.cpp:21]: We might, and it's already in there: `std::vector<Bloock> getChain(const std::string& filename)`

**Comment by ErisMik** [comment]
> One more general comment. I didn't see the client side of peering (I.e pinging all ips looking for initial peers)? Did I miss that somewhere and if so can you point it out?

So this was actually supposed to be a WIP, I thought if I didn't tag you as a reviewer you wouldn't get notifications about it :P

I'll close this since you've already reviewed it, and I'll mark the WIPs better in the future.

**Comment by MuchToKnow** [comment]
> > One more general comment. I didn't see the client side of peering (I.e pinging all ips looking for initial peers)? Did I miss that somewhere and if so can you point it out?
> 
> So this was actually supposed to be a WIP, I thought if I didn't tag you as a reviewer you wouldn't get notifications about it :P
> 
> I'll close this since you've already reviewed it, and I'll mark the WIPs better in the future.

Ah gotcha.  Just thought I should keep up with everything at the very least so I might as well review when I have time. :)

**Pull request merged**